blog/2022-04-25-jwt-vs-api-key-authentication/index.md

```diff
@@ -0,0 +1,109 @@
+---
+title: JWT vs API Key Auth for Machine to Machine APIs
+authors: nate
+tags: [apis, authentication, code]
+date: 2022-04-25
+---
+
+Authorizing other services (i.e. "machines") to call your API is typically done with either JWT tokens or API Keys. The reason to use one or the other varies by your use. This post will explain the pros and cons of each and suggest when each one is a good fit for securing your API.
+
+## JWT Based Machine to Machine Authentication
+
+JWT authentication typically uses an OAuth 2.0 identity provider such as Auth0, AWS Cognito, etc. The identity provider issues tokens after validating the clients are who they say they are.
+
+[![Client Credentials sequence](./auth-sequence-client-credentials.png)](https://auth0.com/docs/get-started/authentication-and-authorization-flow/client-credentials-flow)
+
+When the client sends a request to the API it includes the JWT in the request's `Authorization` header. The API then validates the JWT to be authentic and can then use the information in the JWT to identify the client. Typically the JWT contains a `sub` parameter that identifies the client. The token also includes a `aud` parameter that specifies which API the token can call.
+
+JWT tokens can be issued with any length of expiration time, but it is typical for tokens to expire in a short period, such as one hour.
+
+JWT auth with OAuth uses the [Client Credentials](https://auth0.com/docs/get-started/authentication-and-authorization-flow/client-credentials-flow) flow on the identity server. Each client that will call the API is issued a `Client Id` and a `Client Secret` - think of these values like a username and password. The client uses these values to request an access token they use to call the API. In code, the client credentials flow looks like the following example.
+
+**Request**
+
+```bash
+curl --request POST \
+  --url 'https://YOUR_DOMAIN/oauth/token' \
+  --header 'content-type: application/x-www-form-urlencoded' \
+  --data grant_type=client_credentials \
+  --data client_id=YOUR_CLIENT_ID \
+  --data client_secret=YOUR_CLIENT_SECRET \
+  --data audience=YOUR_API_IDENTIFIER
+```
+
+**Response**
+
+```json
+{
+  "access_token": "eyJz93a...k4laUWw",
+  "token_type": "Bearer",
+  "expires_in": 86400
+}
+```
+
+### Considerations of Machine-to-Machine JWT Auth
+
+JWT-based API auth is a good choice for securing microservices, implementing services within an organization, or sharing APIs with certain types of external clients.
```

Comment:
didn't Vittorio actually disagree with this?

Reply:
no, completely different context